### PR TITLE
Improve docs and tests for `zircon_object::task`

### PR DESCRIPTION
**Pull request by xxchan**
- #![deny(missing_docs)], but #[allow(missing_docs)] for some data enum/structs(that's great pain to add docs to them...)

- Some minor refactors and fixes.

- Thread running & exception not tested. Need further study.

### CODE REVIEW

**Comment by wangrunji0408** [comment]
I just merged #108 which added lots of code about exceptions.
Could you resolve the conflicts and add necessary docs again?
Sorry for the trouble I caused you...

**Comment by xxchan** [comment]
no problem

**Comment by xxchan** [comment]
@benpigchu I noticed two changes
- killed process/job  will not have signal `JOB_TERMINATED/PROCESS_TERMINATED` ?
- `proc.exit` will not kill `critical_to_job` now.

btw, what's the difference between `kill`, `exit` and `terminate`?

**Comment by benpigchu** [comment]
@xxchan 

In the previous implementation, it is possible that the process is terminated, but one thread of it is still not terminated, which is not the expected behavior.

In the current implementation, task do not terminate immediately when they are killed, they will terminate after all its children are terminated. Specially, killing a thread only make it dying. It will terminate after some cleanups in the thread.

Kill and exit are nearly the same now.

Also, process will kill the critical_job when it is terminated instead of killed.

**Comment by benpigchu** [comment]
@xxchan 

Seems I forget to terminate the job/process that do not have a child task. You can fix it here.

**Comment by xxchan** [comment]
@benpigchu Ok, I'll try to fix that

**Comment by xxchan** [comment]
done. PTAL @benpigchu @wangrunji0408

**Comment by wangrunji0408** [comment]
@benpigchu Any idea about the failed core-test cases?

File: zircon-object/src/task/exception.rs [hunk 1]
```diff
@@ -105,11 +109,11 @@ impl Exceptionate {
...
 
 #[repr(C)]
-pub struct ExceptionInfo {
-    pub pid: KoID,
-    pub tid: KoID,
-    pub type_: ExceptionType,
-    pub padding: u32,
+struct ExceptionInfo {
+    tid: KoID,
+    pid: KoID,
```
Comment (wangrunji0408) [inline zircon-object/src/task/exception.rs:114]: Wrong order. Maybe here is why the core-test failed.

Comment (xxchan) [reply zircon-object/src/task/exception.rs]: Oops, I missed the change when merging :(

**Review by wangrunji0408 — APPROVED** [review]
Great!

**Pull request merged**